build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ParseExternalGitDependencies.cs

```diff
@@ -0,0 +1,56 @@
+﻿
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.BuildTools.PrepTasks
+{
+	public class ParseExternalGitDependencies : Task
+	{
+		[Required]
+		public string ExternalFilePath { get; set; }
+		
+		/* %(ExternalGitDependencies.Owner)     - Repo owner
+		 * %(ExternalGitDependencies.Name)      - Repo name
+		 * %(ExternalGitDependencies.Branch)    - Branch name
+		 * %(ExternalGitDependencies.Identity)  - Commit hash
+		 */
+		[Output]
+		public ITaskItem[] ExternalGitDependencies { get; set; }
+
+		static readonly Regex externalRegex = new Regex (
+			@"^(?<comment>\#\s*)?(?<owner>.*)\/(?<repo>.*):(?<branch>.*)@(?<commit>.*)$",
```

Comment:
I would suggest this regex:

```csharp
new Regex(@"^\s*(?<comment>#.*)|(\s*(?<owner>[^/]+)/(?<repo>[^/:]+)[/:](?<branch>[^@]+)@(?<commit>.*))$");
```

...which should probably be "updated" to use `RegexOptions.IgnorePatternWhitespace`; see also: https://github.com/xamarin/xamarin-android/blob/7234e00176a0167c6ebc8bab8eb5ad47ddd7f1fd/src/Xamarin.Android.Build.Tasks/Tasks/AndroidToolTask.cs#L88-L110

Rationale:

  * Allows comments to have leading whitespace, because why not?
  * Allows owners to also have leading whitespace, which is ignored, because why not?
  * Allows `/` to be used to separate the repo from the branch name, a'la GitHub PR messages.
  * `*` in regex is "greedy", and -- for better or worse -- it's perfectly valid for branch names to contain `/`.  For example, https://github.com/grendello/xamarin-android/tree/xa/bootstrap is the `xa/bootstrap` branch.  
    Additionally, this regex places no restrictions on the commit.  This means you can confuse the regex with a "badly crafted" value:

        grendello/xamarin-android:xa/bootstrap@foo:hash@blorp

    With the original regex, `owner`=`grendello/xamarin-android:xa`, `repo`=`bootstrap@foo`, `branch`=`hash`, and `commit`=`blorp`.

    With the updated regex, we don't allow arbitrary `.*` but instead restrict to particular characters, which gives: `owner`=`grendello`, `repo`=`xamarin-android`, `branch`=`xa/bootstrap`, and `commit`=`foo:hash@blorp`.

    A downside with my version is that `@` is a valid character in branch names; see https://stackoverflow.com/a/3651867/83444, and my regex wouldn't match that.

I'm not sure how to easily solve that, unless we can add restrictions on the "commit", and I'm not sure we actually want to do so (as tag names might be useful constructs).

Reply:
I've partially applied these changes, though the following scenarios will remain unsupported:

* Allowing `/` to be used to separate the repo from the branch name
* Allowing `@` in branch names.

I think these are reasonable limitations for now, given that breaking the schema here would leave us in an incompatible state with various other tools as mentioned previously. We can look to extend this in the future alongside other tooling updates as needed.